### PR TITLE
(#1) Migrate to Epson SDK iOS

### PR DESCRIPTION
**Pull request by mdsiddiq-till**
Epson SDK iOS v2.19.0

## Related Issues

Issue: #XXX

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## Screenshots (if appropriate)

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -28,147 +28,91 @@ export interface INetPrinter {
...
+} => {
+  let timeoutId;
+  const timeoutPromise: Promise<INetPrinter> = new Promise((_, reject) => {
+    timeoutId = setTimeout(() => {
+      reject(new Error('Request timed out'));
+    }, SDK_RESPONSE_TIMEOUT);
+  });
+  return {
+    promiseOrTimeout: Promise.race([promise, timeoutPromise]),
+    // @ts-ignore
```
Comment (lakshmaji-till) [inline src/index.ts:49]: typo

File: src/index.ts [hunk 2]
```diff
@@ -28,147 +28,91 @@ export interface INetPrinter {
...
-    ),
-
-  connectPrinter: (inner_mac_address: string): Promise<IBLEPrinter> =>
-    new Promise((resolve, reject) =>
-      RNBLEPrinter.connectPrinter(
-        inner_mac_address,
-        (printer: IBLEPrinter) => resolve(printer),
-        (error: Error) => reject(error)
+  connectAndSend: (bdAddress: string, data: Buffer): Promise<INetPrinter> => {
+    const { promiseOrTimeout, timeoutId } = promiseWithTimeout(
```
Comment (lakshmaji-till) [inline src/index.ts:74]: i have this question: dont we need method for ble device discovery!

File: src/index.ts [hunk 3]
```diff
@@ -28,147 +28,91 @@ export interface INetPrinter {
...
-        port,
-        Platform.OS === 'ios' ? data.toString('hex') : data.toString('base64'),
-        (printer: INetPrinter) => resolve(printer),
-        (error: Error) => reject(error)
+    const { promiseOrTimeout, timeoutId } = promiseWithTimeout(
+      new Promise((resolve, reject) =>
+        RNNetPrinter.connectAndSend(
+          host,
+          port,
+          data.toString('base64'),
```
Comment (snehithv) [inline src/index.ts:104]: for iOS we need to send `hex` in the old world.

Will base64 work?

File: src/index.ts (hunk 2 above)
Comment (snehithv) [reply src/index.ts:74]: We will add that in the future.

For now, we add the bdAddress directly

File: ios/utils/NSDataAdditions.h [hunk 4]
```diff
@@ -0,0 +1,10 @@
+//
+//  NSDataAdditions.h
+//
+//  Created by Till POS on 14/09/21.
+//
+
+@interface NSData ()
```
Comment (snehithv) [inline ios/utils/NSDataAdditions.h:7]: What is the purpose of this header and implementation?

File: ios/utils/EpsonUtils.m [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+#import "EpsonUtils.h"
+#import "PrinterSDK.h"
+
+@implementation EpsonUtils
+
++ (NSString *)makeErrorMessage:(Epos2PrinterStatusInfo *)status
+{
+    NSMutableString *errMsg = [[NSMutableString alloc] initWithString:@""];
+
+    if (status.getOnline == EPOS2_FALSE) {
```
Comment (snehithv) [inline ios/utils/EpsonUtils.m:17]: So epson sdk only gives code and then we add formatted message strings?

File: ios/RNBLEPrinter.h [hunk 6]
```diff
@@ -15,13 +14,11 @@
 #import <React/RCTBridgeModule.h>
 #endif
 #import <CoreBluetooth/CoreBluetooth.h>
+#import "PrinterSDK.h"
 
 @interface RNBLEPrinter : NSObject <RCTBridgeModule>{
-    NSMutableArray* _printerArray;
-    NSObject* m_printer;
+    Epos2Printer *printer_;
```
Comment (snehithv) [inline ios/RNBLEPrinter.h:20]: why to call it printer_, just call it printer

File: ios/RNBLEPrinter.m [hunk 7]
```diff
@@ -20,90 +27,49 @@ - (dispatch_queue_t)methodQueue
...
-                [mapped addObject:dict];
-            }];
-            NSMutableArray *uniquearray = (NSMutableArray *)[[NSSet setWithArray:mapped] allObjects];;
-            successCallback(@[uniquearray]);
-        }];
-    } @catch (NSException *exception) {
-        errorCallback(@[exception.reason]);
-    }
-}
+        printer_ = [[Epos2Printer alloc] initWithPrinterSeries:1 lang:EPOS2_MODEL_ANK];
```
Comment (snehithv) [inline ios/RNBLEPrinter.m:37]: PrinterSeries is set to 1, what does it mean?

File: ios/RNBLEPrinter.m [hunk 8]
```diff
@@ -20,90 +27,49 @@ - (dispatch_queue_t)methodQueue
...
+            [NSException raise:@"Invalid connection" format:@"Can't connect to printer %@", bdAddress];
         }
+        NSData* payload = [NSData dataWithBase64EncodedString:text];
+        [printer_ addCommand:payload];
+        result = [printer_ sendData:5000];
+        if (result != EPOS2_SUCCESS) {
+            [NSException raise:@"Print failed" format:@"Error occurred while printing"];
+        }
+        _successCallback = successCallback;
+        _errorCallback = errorCallback;
```
Comment (snehithv) [inline ios/RNBLEPrinter.m:53]: do we need to attach errorCallBack for success case too?

File: ios/RNNetPrinter.m [hunk 9]
```diff
@@ -1,190 +1,72 @@
...
-RCT_EXPORT_METHOD(printRawData:(NSString *)text
-                  fail:(RCTResponseSenderBlock)errorCallback) {
-    @try {
-
-        !connected_ip ? [NSException raise:@"Invalid connection" format:@"Can't connect to printer"] : nil;
-
-        [[PrinterSDK defaultPrinterSDK] sendHex:text];
-    } @catch (NSException *exception) {
-        errorCallback(@[exception.reason]);
+- (void) onPtrReceive:(Epos2Printer *)printerObj code:(int)code status:(Epos2PrinterStatusInfo *)status printJobId:(NSString *)printJobId
```
Comment (snehithv) [inline ios/RNNetPrinter.m:58]: where is this `onPtrReceive` invoked from?

File: src/index.ts (hunk 3 above)
Comment (mdsiddiq-till) [reply src/index.ts:104]: Handled it in the next pr. (where both sdks are included)

File: ios/utils/NSDataAdditions.h (hunk 4 above)
Comment (mdsiddiq-till) [reply ios/utils/NSDataAdditions.h:7]: Converting base64 to binary.

File: ios/utils/EpsonUtils.m (hunk 5 above)
Comment (mdsiddiq-till) [reply ios/utils/EpsonUtils.m:17]: Yes

File: ios/RNBLEPrinter.h (hunk 6 above)
Comment (mdsiddiq-till) [reply ios/RNBLEPrinter.h:20]: Updated. in the next pr.

File: ios/RNBLEPrinter.m (hunk 7 above)
Comment (mdsiddiq-till) [reply ios/RNBLEPrinter.m:37]: It has no functionality effects. It is the exact printer model.

File: ios/RNBLEPrinter.m (hunk 8 above)
Comment (mdsiddiq-till) [reply ios/RNBLEPrinter.m:53]: At this point it is not success case

File: ios/RNNetPrinter.m (hunk 9 above)
Comment (mdsiddiq-till) [reply ios/RNNetPrinter.m:58]: We attach this function to printer object above.

**Review by snehithv — APPROVED** [review]

**Pull request merged**